### PR TITLE
Ported tiobench test from autotest to avocado [v2].

### PR DESCRIPTION
**Pull request by rajashreer7**
Added tiobench test for avocado with changes in #29 

Signed-off-by: Rajashree rajashr7@linux.vnet.ibm.com

### CODE REVIEW

File: io/disk/tiobench.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+# Based on code by
+#   Author: Lucas Meneghel Rodrigues <lmr@redhat.com>
+#   Author: Xu Tian <xutian@redhat.com>
+#   Author: Jeff Moyer <jmoyer@redhat.com>
+#   Author: Dmitry Monakhov <dmonakhov@openvz.org>
+#   Copyright: 2008 - 2015 Red Hat, Inc.
+#   Copyright: 2013 Red Hat, Inc.
+#   Copyright: 2010 Red Hat, Inc.
+#   Copyright: 2012 <dmonakhov@openvz.org>
```
Comment (ldoktor) [inline io/disk/tiobench.py:25]: Might I ask you where you got these authors from? In autotest I see:
- client/tests/tiobench/control => `walkinair@cn.ibm.com`
- git log -- tests/tiobench => `Yao Fei Zhu <walkinair@cn.ibm.com>`

There are other contributors when I look in `git log`, but they never claimed authorship, so unless they do we don't have to include them. (GPL says we have to include the original author as well as everybody who claimed the authorship after a significant contribution).

File: io/disk/tiobench.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        :params num_runs: This number specifies over how many runs
+                          each test should be averaged.
+        :params iterations: Number of iterations the test needs to be run.
+        """
+        target = self.params.get('target', default='/var/tmp')
+        blocks = self.params.get('blocks', default=4096)
+        threads = self.params.get('threads', default=10)
+        size = self.params.get('size', default=1024)
+        num_runs = self.params.get('numruns', default=2)
+        self.whiteboard = process.system_output('./tiobench.pl --target {}'
```
Comment (ldoktor) [inline io/disk/tiobench.py]: There is a missing ` `after`--target {}`so the`--block` is part of it's value.

File: io/disk/tiobench.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+                                                '--block={} --threads={}'
```
Comment (ldoktor) [inline io/disk/tiobench.py]: the same story on each line...

File: io/disk/tiobench.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+                                                '--size={} --numruns={}'
+                                                .format(target, blocks,
+                                                        threads, size,
+                                                        num_runs))
+        self.log.info(self.whiteboard)
```
Comment (ldoktor) [inline io/disk/tiobench.py]: This is kind-of duplicate as the `process.system_output` already prints the output, so I'd suggest removing this one (alternatively using `verbose=False` in the `process.system_output`.

File: io/disk/tiobench.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        whiteboard_file = os.path.join(self.logdir, 'whiteboard')
+        genio.write_file(whiteboard_file, self.whiteboard)
```
Comment (ldoktor) [inline io/disk/tiobench.py]: This is not necessary, whiteboard is written by avocado automatically.

File: io/disk/tiobench.py.data/makefile.patch [hunk 6]
```diff
@@ -0,0 +1,19 @@
+--- tiobench-0.3.3/Makefile.orig	2011-02-07 20:32:53.000000000 -0800
```
Comment (ldoktor) [inline io/disk/tiobench.py.data/makefile.patch]: I don't see where this patch is applied.

File: io/disk/tiobench.py.data/tiotest.patch [hunk 7]
```diff
@@ -0,0 +1,22 @@
+diff --git a/tiotest.c b/tiotest.c
```
Comment (ldoktor) [inline io/disk/tiobench.py.data/tiotest.patch]: The same story here, do you want to apply them, or is it just from autotest? I mean the compilation and execution works without it. (this one is actually just style fix, the Makefile fix could be serious if one wants to use custom CC)

**Comment by ldoktor** [comment]
Hello @rajashreer7, thank you it's shaping really great. I noticed you're using git, how about using `asset` instead? The change is minimal and you only download zipped version and additionally it get's cached for the future. I tried that and this seems to work fine:

``` diff
diff --git a/io/disk/tiobench.py b/io/disk/tiobench.py
index 26bcf06..0fdfa36 100644
--- a/io/disk/tiobench.py
+++ b/io/disk/tiobench.py
@@ -29,6 +29,7 @@ import os

 from avocado import Test
 from avocado import main
+from avocado.utils import archive
 from avocado.utils import process
 from avocado.utils import build
 from avocado.utils import git
@@ -57,10 +58,11 @@ class Tiobench(Test):
         s_mngr = SoftwareManager()
         if not s_mngr.check_installed("gcc") and not s_mngr.install("gcc"):
             self.error('Gcc is needed for the test to be run')
-        git.get_repo('https://github.com/mkuoppal/tiobench.git',
-                     destination_dir=self.srcdir)
-        os.chdir(self.srcdir)
-        build.make(self.srcdir)
+        locations = ["https://github.com/mkuoppal/tiobench/archive/master.zip"]
+        tarball = self.fetch_asset("tiobench.zip", locations=locations)
+        archive.extract(tarball, self.srcdir)
+        os.chdir(os.path.join(self.srcdir, "tiobench-master"))
+        build.make(".")

     def test(self):
         """
@@ -79,15 +81,13 @@ class Tiobench(Test):
         threads = self.params.get('threads', default=10)
         size = self.params.get('size', default=1024)
         num_runs = self.params.get('numruns', default=2)
-        self.whiteboard = process.system_output('./tiobench.pl --target {}'
-                                                '--block={} --threads={}'
+        self.whiteboard = process.system_output('perl ./tiobench.pl --target {} '
+                                                '--block={} --threads={} '
                                                 '--size={} --numruns={}'
                                                 .format(target, blocks,
                                                         threads, size,
                                                         num_runs))
-        self.log.info(self.whiteboard)
-        whiteboard_file = os.path.join(self.logdir, 'whiteboard')
-        genio.write_file(whiteboard_file, self.whiteboard)
+

 if __name__ == "__main__":
     main()
```

(you might need to polish it a bit, or feel free to ignore it if you have reasons for using git)

File: io/disk/tiobench.py.data/tiobench.yaml [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+            blocks: 4096
+        second:
+            blocks: 8192
+
+    thread: !mux
+        default:
+            threads: 10
+
+    blocksize: !mux
+        default:
```
Comment (ldoktor) [inline io/disk/tiobench.py.data/tiobench.yaml]: there are lots of empty characters in this file, please remove them. (for example at the end of this line)

File: io/disk/tiobench.py.data/README.txt [hunk 9]
```diff
@@ -0,0 +1,13 @@
+This program runs  a multi-threaded I/O benchmark test to measure file system performance in four basic operations:
```
Comment (ldoktor) [inline io/disk/tiobench.py.data/README.txt]: Do you insist on having this extra file here? I'd prefer putting this info as header docstring of the test itself, which is the first thing people see when opening the test. Additionally there are efforts in providing the test's docstring on dry-runs/inquire (not yet settled, but the idea is to have reliable source of information for each test without the need to opening the sources).

Anyway it's not a strong demand, perhaps a question on how are you using this info. Our QA uses the docstring to learn about the test.

File: io/disk/tiobench.py.data/tiobench.yaml [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+        default: 
+            size: 1024
+     
+    runs: !mux
+        default: 
+            numruns: 2
+  
+    iteration: !mux
+        default:
+            iterations: 5
```
Comment (ldoktor) [inline io/disk/tiobench.py.data/tiobench.yaml]: The `iterations` itself won't work. Either you have to list the iterations along with `!mux`:

``` yaml
    iterations: !mux
        default: !mux
            1:
            2:
            3:
            4:
            5:
```

or you need to implement the iterations inside the test.

I'm actually wondering why do you need iterations, I'd just increase the `numruns` if I wanted multiple results. What is the benefit in executing the test multiple times?

File: io/disk/tiobench.py [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+        if not s_mngr.check_installed("gcc") and not s_mngr.install("gcc"):
+            self.error('Gcc is needed for the test to be run')
+        git.get_repo('https://github.com/mkuoppal/tiobench.git',
+                     destination_dir=self.srcdir)
+        os.chdir(self.srcdir)
+        build.make(self.srcdir)
+
+    def test(self):
+        """
+        Runs two iterations of test.
```
Comment (ldoktor) [inline io/disk/tiobench.py]: Really? I think it runs `tiobench` test once.

File: io/disk/tiobench.py [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+        :params target: The directory in which to test.
+                        Defaults to ., the current directory.
+        :params blocks: The blocksize in Bytes to use. Defaults to 4096.
+        :params threads: The number of concurrent test threads.
+        :params size: The total size in MBytes of the files may use together.
+        :params num_runs: This number specifies over how many runs
+                          each test should be averaged.
+        :params iterations: Number of iterations the test needs to be run.
```
Comment (ldoktor) [inline io/disk/tiobench.py]: This param is not used.

File: io/disk/tiobench.py.data/tiobench.yaml (hunk 10 above)
Comment (ldoktor) [inline io/disk/tiobench.py.data/tiobench.yaml]: Btw you can verify the result by `avocado multiplex io/disk/tiobench.py.data/tiobench.yaml` (use `-h` to see options).

**Comment by rajashreer7** [comment]
@ldoktor : Sure, will make those changes.

**Comment by rajashreer7** [comment]
Hello @ldoktor, I have made those changes which were told. Please review it and let me know if there are any more changes to be made.

File: io/disk/tiobench.py (hunk 1 above)
Comment (ldoktor) [inline io/disk/tiobench.py:25]: I still don't see the original author in this license.

File: io/disk/tiobench.py [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+from avocado import main
+from avocado.utils import archive
+from avocado.utils import build
+from avocado.utils import process
+from avocado.utils.software_manager import SoftwareManager
+
+
+class Tiobench(Test):
+    """
+    Avocado module for tiobench.
```
Comment (ldoktor) [inline io/disk/tiobench.py:45]: Avocado test for tiobench

File: io/disk/tiobench.py [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+        Test execution with necessary arguments.
+        :params target: The directory in which to test.
+                        Defaults to ., the current directory.
+        :params blocks: The blocksize in Bytes to use. Defaults to 4096.
+        :params threads: The number of concurrent test threads.
+        :params size: The total size in MBytes of the files may use together.
+        :params num_runs: This number specifies over how many runs
+                          each test should be averaged.
+        """
+        target = self.params.get('target', default='/var/tmp')
```
Comment (ldoktor) [inline io/disk/tiobench.py:73]: Please use `self.workdir` to avoid clashes when running in parallel.

Comment (ldoktor) [inline io/disk/tiobench.py:73]: Actually I'd even change this to:

``` python
target = self.params.get('target') or self.workdir
```

To handle invalid input gracefully, what do you think? (eg. when one specifies `target: None` in the yaml file)

**Comment by ldoktor** [comment]
Hello @rajashreer7 it's almost there. The license is still incorrect as the original author is someone else (you need to look at the original location, which is inside `autotest` and not `autotest-client-tests`. Only the first committer listed himself in the `autotest's` license, so we should reflect that (unless they claim their contribution).

Please take care of that, the tmp location and the docstring and we're ready to go.

File: io/disk/tiobench.py (hunk 1 above)
Comment (rajashreer7) [inline io/disk/tiobench.py:25]: I couldn't find the author name Yao Fei Zhu walkinair@cn.ibm.com in git log and no copyright details are found. Can you tell me from where did you get the git log?

**Pull request closed without merging**